ipie/addons/adafqmc/utils/miscellaneous.py

```diff
@@ -0,0 +1,316 @@
+        return trial
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        (tangent,) = ctx.saved_tensors  # Retrieve precomputed gradient
+        return (
+            torch.einsum("abij, ab -> ij", tangent, grad_output),
+            None,
+            None,
+        )  # None for precomputed_grad's grad
+
+
+def trial_tangent(coupling, trial, tangent):
+    return TrialwithTangent.apply(coupling, trial, tangent)
+
+
+def trial_tangent_matcoupling(coupling, trial, tangent):
+    return TrialwithTangentMatCoupling.apply(coupling, trial, tangent)
+
+
+def dump_hamiltonian(ham, filename):
+    """Dump Hamiltonian to a HDF5 file
+
+    Parameters
+    ----------
+    ham : HamObs object
+        The Hamiltonian with the observable coupled
+    filename : str
+        Path to the HDF5 file to save the Hamiltonian
+    """
+    with h5py.File(filename, "w") as f:
+        f.create_dataset("hcore", data=ham.h1e)
+        f.create_dataset("chol", data=ham.chol)
+        f.create_dataset("nuc", data=ham.enuc)
+        f.create_dataset("packedchol", data=ham.packedchol)
+        f.create_dataset("nelec0", data=ham.nelec0)
+        if ham.obs is not None:
+            f.create_dataset("observable_mat", data=ham.obs[0])
+            f.create_dataset("observable_const", data=ham.obs[1])
+    return
+
+
+def read_hamiltonian_from_h5(filename, obs_type="dipole"):
+    """Read Hamiltonian from a HDF5 file
+
+    Parameters
+    ----------
+    filename : str
+        Path to the HDF5 file containing the Hamiltonian
+    Returns
+    -------
+    ham : HamObs object
+        The Hamiltonian with the observable coupled
+    """
+    with h5py.File(filename, "r") as f:
+        hcore = torch.from_numpy(f["hcore"][()])
+        chol = torch.from_numpy(f["chol"][()])
+        nuc = torch.from_numpy(f["nuc"][()])
+        packedchol = torch.from_numpy(f["packedchol"][()])
+        nelec0 = np.asarray(f["nelec0"]).item()
+        nao = hcore.shape[0]
+        if "observable_mat" in f:
+            obsmat = torch.from_numpy(f["observable_mat"][()])
+            obsconst = torch.from_numpy(f["observable_const"][()])
+            obs = (obsmat, obsconst)
+        else:
+            obs = None
+    ham = HamObs(nelec0, nao, hcore, chol, nuc, obs, packedchol, obs_type)
+    return ham
+
+
+def remove_outliers(data):
+    """Remove outliers for a given list of data using MAD (Median Absolute Deviation)
+
+    Parameters
+    ----------
+    data : numpy.ndarray
+        List of data
+    Returns
+    -------
+    numpy.ndarray
+        List of data with outliers removed
+    """
+    data = data.real
+    assert data.ndim == 1
+    median = np.median(data)
+    b = 1.4826
+    mad = b * np.median(np.abs(data - median))
+    # Remove data points that are more than 3 MADs from the median
+    indices = np.where(np.abs(data - median) <= 10 * mad)[0]
+    return indices
+
+
+def str_to_bool(val):
```

Comment:
Do you still need this function? Delete if not.

Reply:
No, I don't. I'll delete it